File: Atlassian/Karat/snake-paths/typescript/snakes-paths.ts
import { board1, board2, board3, board4, board5 } from "./data";
import { Board, PassableLanes } from "./types";




function findPassableLanes(board: Board): PassableLanes {
  const numRows = board.length;

  const passableLanes: PassableLanes = {
    rows: [],
    cols: [],
  };

  if (numRows === 0) {
    return passableLanes;
  }

  const numCols = board[0].length;

  // find PassableRows
  for (let i = 0; i < numRows; i++) {
    let pass = true;
    for (let j = 0; j < numCols; j++) {
      if (board[i][j] == '+') {
        pass = false;
        break;
      }
    }
    if (pass === true) {
      passableLanes.rows.push(i);
    }
  }

  // find passable cols
  for (let j = 0; j < numCols; j++) {
    let pass = true;
    for (let i = 0; i < numRows; i++) {
      if (board[i][j] === '+') {
        pass = false;
        break;
      }
    }

    if (pass === true) {
      passableLanes.cols.push(j);
    }
  }

  return passableLanes;
}

console.log(findPassableLanes(board1));
console.log(findPassableLanes(board2));
console.log(findPassableLanes(board3));
console.log(findPassableLanes(board4));
console.log(findPassableLanes(board5));

File: Atlassian/Karat/snake-paths/go/snakes-paths.go
package main

import "fmt"

type Square byte

type Board [][]Square

type PassableLanes struct {
	rows []int
	cols []int
}

var board1 = Board{
	{'+', '+', '+', '0', '+', '0', '0'},
	{'0', '0', '+', '0', '0', '0', '0'},
	{'0', '0', '0', '0', '+', '0', '0'},
	{'+', '+', '+', '0', '0', '+', '0'},
	{'0', '0', '0', '0', '0', '0', '0'},
}

var board2 = Board{
	{'+', '+', '+', '0', '+', '0', '0'},
	{'0', '0', '0', '0', '0', '+', '0'},
	{'0', '0', '+', '0', '0', '0', '0'},
	{'0', '0', '0', '0', '+', '0', '0'},
	{'+', '+', '+', '0', '0', '0', '+'},
}

var board3 = Board{
	{'+', '+', '+', '0', '+', '0', '0'},
	{'0', '0', '0', '0', '0', '0', '0'},
	{'0', '0', '+', '+', '0', '+', '0'},
	{'0', '0', '0', '0', '+', '0', '0'},
	{'+', '+', '+', '0', '0', '0', '+'},
}

var board4 = Board{{'+'}}
var board5 = Board{{'0'}}

func (board Board) findPassableLanes() PassableLanes {
	numRows := len(board)

	var passableLanes = PassableLanes{
		rows: []int{},
		cols: []int{},
	}

	if numRows == 0 {
		return passableLanes
	}

	numCols := len(board[0])

	// find PassableRows
	for i := 0; i < numRows; i++ {
		pass := true
		for j := 0; j < numCols; j++ {
			if board[i][j] == '+' {
				pass = false
				break
			}
		}
		if pass == true {
			passableLanes.rows = append(passableLanes.rows, i)
		}
	}

	// find passable cols
	for j := 0; j < numCols; j++ {
		pass := true
		for i := 0; i < numRows; i++ {
			if board[i][j] == '+' {
				pass = false
				break
			}
		}

		if pass == true {
			passableLanes.cols = append(passableLanes.cols, j)
		}
	}

	return passableLanes
}

func main() {
	fmt.Printf("%+v\n", board1.findPassableLanes())
	fmt.Printf("%+v\n", board2.findPassableLanes())
	fmt.Printf("%+v\n", board3.findPassableLanes())
	fmt.Printf("%+v\n", board4.findPassableLanes())
	fmt.Printf("%+v\n", board5.findPassableLanes())
}
